src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -45,43 +47,52 @@ object Ledger extends Logger {
-            val refundGasFn = {
-              if (theResult.gasRemaining > 0) (refundGas _).curried(Address(block.header.beneficiary))(gasUsed)
-              else identity[InMemoryWorldStateProxy] _
-            }
-            val payBeneficiariesFn = (payForGasUsedToBeneficiary _).curried(stx)
-            val deleteAccountsFn = (deleteAccounts _).curried(theResult.addressesToDelete)
-            val persistStateFn = InMemoryWorldStateProxy.persistState _
-            val (newWorldStateProxy, newAcumGas) =
-              (refundGasFn andThen payBeneficiariesFn andThen deleteAccountsFn andThen persistStateFn) (theResult.world) -> (gasUsed + acumGas)
+    val parentStateRoot = blockchain.getBlockHeaderByHash(block.header.parentHash).map(_.stateRoot)
+    val initialWorld = InMemoryWorldStateProxy(storages, stateStorage, parentStateRoot)
+
+    block.body.transactionList.foldLeft[BlockResult](BlockResult(worldState = initialWorld)) {
+      case (BlockResult(world, acumGas, receipts), stx)=>
+        validateTransaction(stx, world, acumGas, block.header) match {
+          case Left(err) =>
+            throw new RuntimeException(err)
+
+          case Right(_) =>
+            val TxResult(newWorld, gasUsed, logs) = executeTransaction(stx, block.header, world)
+
             val receipt = Receipt(
-              postTransactionStateHash = newWorldStateProxy.stateRootHash,
-              cumulativeGasUsed = newAcumGas,
-              logsBloomFilter = BloomFilter.create(theResult.logs.toSet),
-              logs = theResult.logs
+              postTransactionStateHash = newWorld.stateRootHash,
+              cumulativeGasUsed = acumGas + gasUsed,
+              logsBloomFilter = BloomFilter.create(logs),
+              logs = logs
             )
-            ExecResult(newWorldStateProxy, newAcumGas, receipts :+ receipt)
-          case Left(error) => throw new RuntimeException(error)
+
+            BlockResult(newWorld, receipt.cumulativeGasUsed, receipts :+ receipt)
         }
     }
   }
 
+  private[ledger] def executeTransaction(stx: SignedTransaction, blockHeader: BlockHeader, world: InMemoryWorldStateProxy): TxResult = {
+    val gasPrice = UInt256(stx.tx.gasPrice)
+    val gasLimit = UInt256(stx.tx.gasLimit)
+
+    val world1 = updateSenderAccountBeforeExecution(stx, world)
+    val result = runVM(stx, blockHeader, world1)
+
+    val gasUsed = if(result.error.isDefined) gasLimit else gasLimit - result.gasRemaining
+    val gasRefund = calcGasRefund(stx, result)
+
+    val refundGasFn = pay(stx.senderAddress, gasRefund * gasPrice) _
+    val payMinerForGasFn = pay(Address(blockHeader.beneficiary), (gasLimit - gasRefund) * gasPrice) _
+    val deleteAccountsFn = deleteAccounts(result.addressesToDelete) _
+    val persistStateFn = InMemoryWorldStateProxy.persistState _
+
+    val world2 = (refundGasFn andThen payMinerForGasFn andThen deleteAccountsFn andThen persistStateFn)(result.world)
```

Comment:
@ntallar you asked: 

> Is it okay for accounts to be deleted in case the execution resulted in an error? At least EthereumJ seems to clear the list of accounts to be deleted when an exception ocurrs.

You're right, but let's make that part of EC-151